libraries/botbuilder-ai-orchestrator/tests/recognizerTelemetryUtils.js

```diff
@@ -0,0 +1,81 @@
+const { ok, strictEqual } = require('assert');
+const sinon = require('sinon');
+const { NullTelemetryClient } = require('botbuilder-core');
+const { BoolExpression } = require('adaptive-expressions');
+const { asMessageActivity } = require('botframework-schema').ActivityEx;
+
+const orchestratorIntentText = 'Hello, Orc!';
+
+const spyOnTelemetryClientTrackEvent = (recognizer) => {
+    const telemetryClient = new NullTelemetryClient();
+    const spy = sinon.spy(telemetryClient, 'trackEvent');
+    recognizer.telemetryClient = telemetryClient;
+
+    return spy;
+};
+
+const getLogPersonalInformation = (recognizer, dialogContext) => {
+    return recognizer.logPersonalInformation instanceof BoolExpression
+        ? recognizer.logPersonalInformation.getValue(dialogContext.state)
+        : recognizer.logPersonalInformation;
+};
+
+const validateTelemetry = async ({ recognizer, dialogContext, spy, activity, result, callCount }) => {
+    const logPii = getLogPersonalInformation(recognizer, dialogContext);
+    const expectedTelemetryProps = getExpectedProps(activity, result, logPii);
+    const actualTelemetryProps = spy.getCall(callCount - 1).args[0];
+
+    strictEqual(spy.callCount, callCount);
+    strictEqual(actualTelemetryProps.name, `${recognizer.constructor.name}Result`);
+    ok(hasValidTelemetryProps(actualTelemetryProps.properties, expectedTelemetryProps, activity));
+};
+
+module.exports = {
+    orchestratorIntentText,
+    getLogPersonalInformation,
+    spyOnTelemetryClientTrackEvent,
+    validateTelemetry,
+};
+
+// **** PRIVATE **** //
+
+const getOrchestratorIntentProps = () => {
+    return {
+        TopIntent: 'mockLabel',
+        TopIntentScore: '0.9',
+        Intents: '{\"mockLabel\":{\"score\":0.9}}',
+        Entities: '{}',
+        AdditionalProperties:
+            '{\"result\":[{\"score\":0.9,\"label\":{\"name\":\"mockLabel\"}}]}',
+    };
+};
+
+const getExpectedProps = (activity, result, logPersonalInformation) => {
+    const text = asMessageActivity(activity).text;
+    const expectedProps = text === orchestratorIntentText ? getOrchestratorIntentProps() : {};
+
+    if (logPersonalInformation) {
+        expectedProps['Text'] = text;
+        expectedProps['AlteredText'] = result.alteredText;
+    }
+
+    return expectedProps;
+};
+
+const hasValidTelemetryProps = (actual, expected, activity) => {
+    if (Object.keys(actual).length !== Object.keys(expected).length) {
+        return false;
+    }
+
+    for (const property in actual) {
```

Comment:
`for...in` can be a bit subtle so I prefer to use `for...of` unless you have a good reason not to. I think this entire loop could be expressed more succinctly as:

```javascript
return Object.entries(actual).every(([key, value]) => expected[key] === value);
```

Reply:
Hmm...there was some issue that I encountered when writing the C# side with the mac pipeline tests that made me do this whole dance around making a whole different check for entities (made test more lenient by checking only that it had the right entity detected, but not checking the value)

Seems like combination of your suggestion of `JSON.stringify(<actual telemetry property obj>)` + `.every(...)` suggestion makes it so that there isn't any workaround work needed for entities! 🎊 

Changed tests to use `return Object.entries(actual).every(([key, value]) => expected[key] === value);` and it's a lot cleaner now